src/Nethermind/Nethermind.Evm/TransactionProcessing/TransactionProcessor.cs

```diff
@@ -93,278 +94,396 @@ public void Trace(Transaction transaction, BlockHeader block, ITxTracer txTracer
+                    return false;
                 }
 
-                if (!noValidation && spec.IsEip1559Enabled && !transaction.IsFree() &&
-                    senderBalance < (UInt256)transaction.GasLimit * transaction.MaxFeePerGas + value)
+                UInt256 senderBalance = _worldState.GetBalance(tx.SenderAddress);
+                if (UInt256.SubtractUnderflow(senderBalance, tx.Value, out UInt256 balanceLeft))
                 {
-                    TraceLogInvalidTx(transaction,
-                        $"INSUFFICIENT_MAX_FEE_PER_GAS_FOR_SENDER_BALANCE: ({caller})_BALANCE = {senderBalance}, MAX_FEE_PER_GAS: {transaction.MaxFeePerGas}");
-                    QuickFail(transaction, block, txTracer, eip658NotEnabled,
-                        "insufficient MaxFeePerGas for sender balance");
-                    return;
+                    TraceLogInvalidTx(tx, $"INSUFFICIENT_SENDER_BALANCE: ({tx.SenderAddress})_BALANCE = {senderBalance}");
+                    QuickFail(tx, blk, spec, tracer, "insufficient sender balance");
+                    return false;
                 }
 
-                if (transaction.Nonce != _worldState.GetNonce(caller))
+                bool overflows = UInt256.MultiplyOverflow((UInt256)tx.GasLimit, tx.MaxFeePerGas, out UInt256 maxGasFee);
+                if (spec.IsEip1559Enabled && !tx.IsFree() && (overflows || balanceLeft < maxGasFee))
                 {
-                    TraceLogInvalidTx(transaction,
-                        $"WRONG_TRANSACTION_NONCE: {transaction.Nonce} (expected {_worldState.GetNonce(caller)})");
-                    QuickFail(transaction, block, txTracer, eip658NotEnabled, "wrong transaction nonce");
-                    return;
+                    TraceLogInvalidTx(tx, $"INSUFFICIENT_MAX_FEE_PER_GAS_FOR_SENDER_BALANCE: ({tx.SenderAddress})_BALANCE = {senderBalance}, MAX_FEE_PER_GAS: {tx.MaxFeePerGas}");
+                    QuickFail(tx, blk, spec, tracer, "insufficient MaxFeePerGas for sender balance");
+                    return false;
                 }
 
-                _worldState.IncrementNonce(caller);
+                overflows = UInt256.MultiplyOverflow((UInt256)tx.GasLimit, effectiveGasPrice, out senderReservedGasPayment);
+                if (overflows || senderReservedGasPayment > balanceLeft)
+                {
+                    TraceLogInvalidTx(tx, $"INSUFFICIENT_SENDER_BALANCE: ({tx.SenderAddress})_BALANCE = {senderBalance}");
+                    QuickFail(tx, blk, spec, tracer, "insufficient sender balance");
+                    return false;
+                }
             }
 
-            // Do not charge gas if noValidation is set
-            if (!noValidation) _worldState.SubtractFromBalance(caller, senderReservedGasPayment, spec);
-            if (commit)
+            if (validate)
+                _worldState.SubtractFromBalance(tx.SenderAddress, senderReservedGasPayment, spec);
```

Comment:
I think we can move it inside if, as senderReservedGasPayment for system txs is zero